cmd/machine-config-controller/start.go

```diff
@@ -172,6 +172,7 @@ func createControllers(ctx *ctrlcommon.ControllerContext) []ctrlcommon.Controlle
 			ctx.ConfigInformerFactory.Config().V1().Images(),
 			ctx.ConfigInformerFactory.Config().V1().ImageDigestMirrorSets(),
 			ctx.ConfigInformerFactory.Config().V1().ImageTagMirrorSets(),
+			ctx.ConfigInformerFactory.Config().V1alpha1().ClusterImagePolicies(),
```

Comment:
@JoelSpeed  This seems ok from the ci [logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4160/pull-ci-openshift-machine-config-operator-master-e2e-aws-ovn/1757843506316972032/artifacts/e2e-aws-ovn/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-controller-5c47b7b49d-nhdjk_machine-config-controller.log)? around `Starting MachineConfigController-ContainerRuntimeConfigController`, there's no accessing API error

Reply:
```
W0214 19:33:25.385128       1 reflector.go:539] github.com/openshift/client-go/config/informers/externalversions/factory.go:125: failed to list *v1alpha1.ClusterImagePolicy: the server could not find the requested resource (get clusterimagepolicies.config.openshift.io)
E0214 19:33:25.385182       1 reflector.go:147] github.com/openshift/client-go/config/informers/externalversions/factory.go:125: Failed to watch *v1alpha1.ClusterImagePolicy: failed to list *v1alpha1.ClusterImagePolicy: the server could not find the requested resource (get clusterimagepolicies.config.openshift.io)
```
This is repeated every 45 seconds through the logs showing the informer is having issues starting

Reply:
> You only want to add this if the feature gate is enabled, else the informer when it starts will try to access an API type that does not exist

Tried to fix this review by commit  [2cc6e6d](https://github.com/openshift/machine-config-operator/commit/2cc6e6d79db828f14d5d932100aa721550f73eae)
The featuregete will have a value only when the `(*Controller) Run()` is called (featuregate from [start.go](https://github.com/openshift/machine-config-operator/blob/master/cmd/machine-config-controller/start.go)). So I added the `ctrl.enableSigstoreAPIHandler()` inside the `(*Controller) Run()` . 
But The cluster failed to launch with techpreview turned on. e2e-gcp-op-techpreview [logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4160/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op-techpreview/1758258587399884800/artifacts/e2e-gcp-op-techpreview/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-controller-76f6dbbf5-mvb8j_machine-config-controller.log) did not log `Starting MachineConfigController-ContainerRuntimeConfigController` after the featuregate observed as true. The containerruntimeconfig controller did not start. maybe blocked on L234 `cache.WaitForCacheSync` https://github.com/openshift/machine-config-operator/blob/2cc6e6d79db828f14d5d932100aa721550f73eae/pkg/controller/container-runtime-config/container_runtime_config_controller.go#L230-L238
If not add `listerCaches = append(listerCaches, ctrl.clusterImagePolicyListerSynced)`, 
the cluster can launch, but the controller did not react to clusterimagepolicy objects created,  it seems the `ctrl.configInformerFactory.Config().V1alpha1().ClusterImagePolicies().Informer().AddEventHandler` did not add listener, but this call to `AddEventHandler` did not return an error.

Reply:
@JoelSpeed  From the finding above https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1506704089, I am not sure how to add listener for the ClusterImagePolicy after the featuregate is enabled. Is it because I can not call AddEventHandler after the sharedInformer is [running](https://github.com/openshift/machine-config-operator/blob/398c1710f12622e9c0f94f99b12abeabed3ac5f7/cmd/machine-config-controller/start.go#L90)?

Reply:
You should know from startup time whether the feature gate is enabled or not, the featuregateaccessor should be synced before you even get to registering your listeners, so, this should be a simple conditional logic around the listener initialisation and event handler adding

Reply:
Currently in the [start.go](https://github.com/openshift/machine-config-operator/blob/master/cmd/machine-config-controller/start.go), all controllers added listeners before detecting the featuregate.
I found I have to start the SharedInformerFactory again in the container runtime config controller `Run()` to let the AddEventHandler take effect.